From e2191ad3ff2e40b83a31e4683c940df6ac09ba41 Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Thu, 16 Jan 2014 15:31:00 +0800
Subject: [PATCH] cpuidle: add unstable tsc condition

In some new intel cpu of x86-64, it has a stable tsc with nonstop_tsc flag.
If writing its tsc, it cause tsc so big that system timeout and soft lockup.

And some x86-64 bsp has HAVE_UNSYNCHRONIZED_TSC definition,
but in fact, it still has stable tsc.
So, we believe the result of soft detecting instead.

Signed-off-by: Li Wang <li.wang@windriver.com>
---
 drivers/cpuidle/cpuidle.c |   15 +++++++++------
 1 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/drivers/cpuidle/cpuidle.c b/drivers/cpuidle/cpuidle.c
index 1c9f85e..f6692d0 100644
--- a/drivers/cpuidle/cpuidle.c
+++ b/drivers/cpuidle/cpuidle.c
@@ -53,7 +53,7 @@ static void cpuidle_idle_call(void)
 	struct cpuidle_state *target_state;
 	int next_state;
 #if defined(CONFIG_X86) && defined(CONFIG_HAVE_UNSYNCHRONIZED_TSC)
-	u64 tsc;
+	u64 tsc = 0;
 	unsigned long flags;
 #endif
 
@@ -90,7 +90,8 @@ static void cpuidle_idle_call(void)
 	/* enter the state and update stats */
 	dev->last_state = target_state;
 #if defined(CONFIG_X86) && defined(CONFIG_HAVE_UNSYNCHRONIZED_TSC)
-	tsc = get_cycles();
+	if (check_tsc_unstable())
+		tsc = get_cycles();
 #endif
 	dev->last_residency = target_state->enter(dev, target_state);
 	/* Some Intel CPUs have unstable TSCs.
@@ -98,10 +99,12 @@ static void cpuidle_idle_call(void)
 	 * So, we need to calibrate the TSC when system is idle(C-State).
 	 */
 #if defined(CONFIG_X86) && defined(CONFIG_HAVE_UNSYNCHRONIZED_TSC)
-	tsc += (u64)(dev->last_residency)*(cpu_khz/1000);
-	local_irq_save(flags);
-	write_tsc((u32)tsc, (u32)((u64)tsc >> 32));
-	local_irq_restore(flags);
+	if (check_tsc_unstable()) {
+		tsc += (u64)(dev->last_residency)*(cpu_khz/1000);
+		local_irq_save(flags);
+		write_tsc((u32)tsc, (u32)((u64)tsc >> 32));
+		local_irq_restore(flags);
+	}
 #endif
 	if (dev->last_state)
 		target_state = dev->last_state;
-- 
1.7.0

